### ISSUE TITLE
Locale support

### ISSUE THREAD

**Issue by IBPX** [goatattack/goatattack#26]
As far as I can tell, Goat Attack doesn't support translating to other languages. Is this going to be implemented? This might require #6 to be fixed first.

**Comment by freanux** [comment]
The font system is implemented straightforward right now. This system needs a complete rewrite to support UTF8 characters. After that, translating to other languages can be done.

**Comment by akien-mga** [comment]
Once ready for internationalisation, I would gladly contribute a French translation.

**Comment by freanux** [comment]
Sounds great.

**Comment by freanux** [comment]
@akien-mga: Goat Attack supports i18n now. I embrace your offer, the translation file is the the shared folder. There you find an I18N folder with several translation src files, eg:
https://github.com/goatattack/goatattack/blob/next/src/shared/i18n/I18NFrench.cpp

Pull the newest version (the next branch) to get the actual version.

**Comment by freanux** [comment]
ah btw, i forgot: Goat Attack needs now libtruetype.

**Comment by akien-mga** [comment]
Nice! I'll have a look. Have you considered used gettext or a similar system to handle translations in non-cpp files? gettext PO files are pretty common and there are great tools to work on those.

Also, I'd advise you to remove debugging messages from the translation templates. Messages like `"Could not create window: ${1}"` are typically more relevant to developers than to users, and you don't want users making bug reports with errors in a language you don't understand.

**Comment by IBPX** [comment]
@akien-mga I agree; gettext .po files would be much simpler, both format-wise and in that there are many editors already existing for them.

> Also, I'd advise you to remove debugging messages from the translation templates.

I actually disagree. I think you should be able to tell what a debug message means in your own language.

**Comment by freanux** [comment]
I know GNU gettext. But there are some disadvantages to use it in that game, particularly in a network environment. As GNU gettext uses a `const char *` for an identifier, I decided to use an `int` to identify the i18n-string:
https://github.com/goatattack/goatattack/blob/4304498a6d8e019ff914ff12799f4a1a11dcf7d5/src/shared/src/I18N.cpp#L79-L88 

If a game server sends a text, it only sends that identifier over the net, not the entire string. Have a look here:
https://github.com/goatattack/goatattack/blob/4304498a6d8e019ff914ff12799f4a1a11dcf7d5/src/Engine/src/ClientEvent.cpp#L770-L793
This is the part, where a client gets the message from the server, with optionally its parameters.

During a gameplay, the UDP traffic should be kept as low as possible.

**Comment by freanux** [comment]
@akien-mga I want to ask if there are any news about the french translation. I plan to release Goat Attack in a few days.

**Comment by akien-mga** [comment]
I had forgotten about it, sorry :) Will try to work on the translation in the coming days, maybe Monday evening. @ObaniGemini might be interested in helping too (see https://github.com/goatattack/goatattack/issues/26#issuecomment-256366217).

**Comment by freanux** [comment]
Perfect. Thanks a lot. :)

**Comment by ObaniGemini** [comment]
yep

**Comment by ObaniGemini** [comment]
There really are **many** duplicates in this locale file, and that without counting the debug info which would surely be easier to understand without any translation ^^
